docs/architecture/adr-051-arbitrary-protobuf-ipld-support-scheme.md

```diff
@@ -0,0 +1,228 @@
+# ADR-051: Arbitrary Protobuf IPLD Support Scheme
+
+## Changelog
+
+- Feb 14th, 2022: Initial Draft
+
+## Status
+
+DRAFT
+
+## Abstract
+This ADR describes a generic IPLD content addressing scheme for the arbitrary protobuf types stored in SMT referenced state storage.
+
+## Context
+
+Because all values stored in the state of a ComsosSDK chain are protobuf encoded, we are presented with a unique
+opportunity to provide extremely rich IPLD content-typing information for the arbitrary types stored in
+(referenced from) the SMT in a generic fashion.
+
+### SDK context
+
+The SDK stores values in state storage in a protobuf encoded format. Each module defines their own custom types in
+.proto definitions and these types are registered with a ProtoCodec that manages the marshalling and unmarshalling of these
+types to and from the binary format that is persisted to disk. As such, an indefinite/unbounded number of protobuf types
+need to be supported within the Cosmos ecosystem at large.
```

Comment:
[Also mentioned here](https://github.com/cosmos/cosmos-sdk/pull/11371#discussion_r827402775) but is this true? Modules get access to e.g. KVStore API which deals in raw `[]byte` which AFAIK don't require protobuf encoding. Are you referring to something else?

Reply:
Thanks for the question/feedback! That is correct, nowhere in the SDK is using protobuf encoding of KVStore values strictly enforced/required. But proto and amino (legacy support) are the only codecs supported in https://github.com/cosmos/cosmos-sdk/blob/master/codec/codec.go and, in general, the goal is to move all encoding to protobuf: https://github.com/cosmos/cosmos-sdk/blob/master/docs/core/encoding.md. So, it is not a strict requirement but it is loosely prescribed/standardized.

In any case, it has been decided that this ADR is not appropriate for the SDK so I'm going to close this PR. Instead I will be finalizing the proposal here as part of the ipld schemas for tendermint/cosmos: https://github.com/ipld/ipld/pull/111.

Reply:
Gotcha. But I guess as long as [KVStore interface](https://pkg.go.dev/github.com/cosmos/cosmos-sdk@v0.45.1/store/types#KVStore) provides Get/Set that operate on `[]byte` there's no way to assert an encoding 😬